bitcoin/crawler.go

```diff
@@ -0,0 +1,546 @@
+	}
+
+	// keep track of all unknown connection errors
+	if bitcoinResult.ConnectErrorStr == models.NetErrorUnknown && bitcoinResult.ConnectError != nil {
+		properties["connect_error"] = bitcoinResult.ConnectError.Error()
+	}
+
+	// keep track of all unknown crawl errors
+	if bitcoinResult.ErrorStr == models.NetErrorUnknown && bitcoinResult.Error != nil {
+		properties["crawl_error"] = bitcoinResult.Error.Error()
+	}
+
+	data, err := json.Marshal(properties)
+	if err != nil {
+		log.WithError(err).WithField("properties", properties).Warnln("Could not marshal peer properties")
+	}
+
+	if len(bitcoinResult.ListenAddrs) > 0 {
+		task.AddrInfo.Addr = bitcoinResult.ListenAddrs
+	}
+
+	cr := core.CrawlResult[PeerInfo]{
+		CrawlerID:           c.id,
+		Info:                task,
+		CrawlStartTime:      crawlStart,
+		RoutingTableFromAPI: false,
+		RoutingTable:        bitcoinResult.RoutingTable,
+		Agent:               bitcoinResult.Agent,
+		Protocols:           bitcoinResult.Protocols,
+		ConnectError:        bitcoinResult.ConnectError,
+		ConnectErrorStr:     bitcoinResult.ConnectErrorStr,
+		CrawlError:          bitcoinResult.Error,
+		CrawlErrorStr:       bitcoinResult.ErrorStr,
+		CrawlEndTime:        time.Now(),
+		ConnectStartTime:    bitcoinResult.ConnectStartTime,
+		ConnectEndTime:      bitcoinResult.ConnectEndTime,
+		Properties:          data,
+		LogErrors:           c.cfg.LogErrors,
+	}
+
+	// We've now crawled this peer, so increment
+	c.crawledPeers++
+
+	return cr, nil
+}
+
+func (c *Crawler) PeerProperties(node *AddrInfo) map[string]any {
+	// TODO: to be implemented later
+	properties := map[string]any{}
+
+	properties["NA"] = true
+
+	return properties
+}
+
+type BitcoinResult struct {
+	ConnectStartTime      time.Time
+	ConnectEndTime        time.Time
+	ConnectError          error
+	ConnectErrorStr       string
+	Agent                 string
+	ProtocolVersion       int32
+	Protocols             []string
+	ListenAddrs           []ma.Multiaddr
+	Transport             string // the transport of a successful connection
+	ConnClosedImmediately bool   // whether conn was no error but still unconnected
+	GenTCPAddr            bool   // whether a TCP address was generated
+	Error                 error
+	ErrorStr              string
+	RoutingTable          *core.RoutingTable[PeerInfo]
+}
+
+func (c *Crawler) crawlBitcoin(ctx context.Context, pi PeerInfo) chan BitcoinResult {
+	resultCh := make(chan BitcoinResult)
+
+	go func() {
+		result := BitcoinResult{}
+
+		// sanitize the given addresses like removing UDP-only addresses and
+		// adding corresponding TCP addresses.
+		sanitizedAddrs, generated := sanitizeAddrs(pi.Addrs())
+
+		// keep track if we generated a TCP address to dial
+		result.GenTCPAddr = generated
+
+		addrInfo := peer.AddrInfo{
+			ID:    pi.ID(),
+			Addrs: sanitizedAddrs,
+		}
+
+		var conn net.Conn
+		result.ConnectStartTime = time.Now()
+		conn, result.ConnectError = c.connect(ctx, addrInfo) // use filtered addr list
```

Comment:
The conn needs to be closed when the `ctx` is cancelled. Then `ctrl` + `c` will work again when running a crawl.

If `c.connect` was successful, spawn a new go routine that selects on ctx.Done() and on a channel that's closed when this function returns. Something like

```go
   done := make(chan struct{})
   defer close(done)
   
   go func() {
      select {
      case <-ctx.Done():
          conn.Close()
      case <-done:
      }
   }()
```